hub/hub.go

```diff
@@ -232,114 +223,175 @@ func (h *ForkableHub) SourceThroughCursor(startBlock uint64, cursor *bstream.Cur
 	return
 }
 
-func (h *ForkableHub) bootstrap(blk *pbbstream.Block) error {
-	zlog.Info("bootstrapping ForkableHub", zap.Stringer("blk", blk.AsRef()))
+func (h *ForkableHub) bootstrap() error {
+	ctx := context.Background()
 
-	// don't try bootstrapping from one-block-files if we are not at HEAD
-	if blk.Number < h.forkable.HeadNum() {
-		zlog.Info("skip bootstrapping ForkableHub from one-block-files", zap.Stringer("blk", blk.AsRef()), zap.Uint64("forkdb_head_num", h.forkable.HeadNum()))
-		return h.forkable.ProcessBlock(blk, nil)
+	sortedOneBlocksFiles, err := h.WalkOneBlocksStore(ctx)
+	if err != nil {
+		return fmt.Errorf("walking through one blocks files: %w", err)
 	}
 
-	if !h.forkable.Linkable(blk) {
-		startBlock := substractAndRoundDownBlocks(blk.LibNum, uint64(h.keepFinalBlocks))
-		zlog.Info("bootstrapping on un-linkable block", zap.Uint64("start_block", startBlock), zap.Stringer("head_block", blk.AsRef()))
+	if len(sortedOneBlocksFiles) == 0 {
+		return fmt.Errorf("no one blocks found")
+	}
 
-		var oneBlocksSource bstream.Source
-		if h.oneBlocksSourceFactoryWithSkipFunc != nil {
-			skipFunc := func(idSuffix string) bool {
-				return h.MatchSuffix(idSuffix)
-			}
-			oneBlocksSource = h.oneBlocksSourceFactoryWithSkipFunc(startBlock, h.forkable, skipFunc)
-		} else {
-			oneBlocksSource = h.oneBlocksSourceFactory(startBlock, h.forkable)
+	mostRecentOneBlock := sortedOneBlocksFiles[len(sortedOneBlocksFiles)-1]
+
+	_, _, _, libNumAsRef, _, err := bstream.ParseFilename(mostRecentOneBlock)
+	if err != nil {
+		return fmt.Errorf("parsing filename: %w", err)
+	}
+
+	oneBlocksUpToLibRef := make([]*pbbstream.Block, 0)
+	for _, filename := range sortedOneBlocksFiles {
+		blockNum, suffixID, _, _, _, err := bstream.ParseFilename(filename)
+		if err != nil {
+			return fmt.Errorf("parsing filename: %w", err)
 		}
 
-		if oneBlocksSource == nil {
-			zlog.Debug("no oneBlocksSource from factory, not bootstrapping hub yet")
-			return nil
+		if blockNum < libNumAsRef {
+			continue
 		}
-		zlog.Info("bootstrapping ForkableHub from one-block-files", zap.Uint64("start_block", startBlock), zap.Stringer("head_block", blk.AsRef()))
-		go oneBlocksSource.Run()
-		select {
-		case <-oneBlocksSource.Terminating():
-			break
-		case <-h.Terminating():
-			return h.Err()
+
+		if availableBlock := h.forkable.GetBlockByHashSuffix(suffixID); availableBlock != nil {
```

Comment:
we could double check the block number too, availableBlock.Number == blockNum
this way we add guarantee that this is the same block, since we have a partial ID.